### PR TITLE
[WIP] ENH: faster faasta (for TV-l1)

### PR DESCRIPTION
**Pull request by GaelVaroquaux**
Careful timing / profiling shows that the dual-gap of the proximal
operator is sometimes decreased too far. As a result, the proximal
operator spends a lot of time trying to reach this dual gap, for no good
increase in overall energy. Profiling the code shows that this is the
number one loss of time.

Given that with warm-restart well implemented, decreasing on the fly the
tolerance of the proximal isn't very costly, we can rely fully on the
adaptative strategy, rather than on the paper of Schmidt et al (ie
implement FAASTA as in the paper). I have done this here. Printing values
of energies on the Haxby dataset shows that energy is decreased as well.

The code seems to be around 1.5 times faster.

I am sending a PR to see how this behaves on Travis / Circle

Cc @dohmatob , @eickenberg : please have a look

### CODE REVIEW

File: nilearn/decoding/fista.py [hunk 1]
```diff
@@ -184,7 +187,9 @@ def mfista(f1_grad, f2_prox, total_energy, lipschitz_constant, w_size,
 
         # backward (prox) step
         for _ in range(10):
-            w, prox_info = f2_prox(z - stepsize * gradient_buffer, stepsize,
+            # z <- -stepsize * gradient_buffer + z
+            z = inplace_add_mult(gradient_buffer, z, a=-stepsize)
```
Comment (eickenberg) [inline nilearn/decoding/fista.py]: I recall from the ADMM version that this assignment to `z` was necessary there, too. If this operation really is in place, it wouldn't need the assignment to `z`. Does it work without it?

File: nilearn/decoding/fista.py [hunk 2]
```diff
@@ -231,7 +236,7 @@ def mfista(f1_grad, f2_prox, total_energy, lipschitz_constant, w_size,
             #
             # For this reason, we rely more on the linesearch-like
             # strategy to set the dgap_tol
-            dgap_tol = abs(energy_delta) / (i + 1.)
+            dgap_tol = abs(energy_delta)
```
Comment (eickenberg) [inline nilearn/decoding/fista.py]: makes sense.

While we're at it, should we think about relaxing the dual gap tolerance again by line searching over it with warm starts? Probably for another PR.

File: nilearn/decoding/fista.py (hunk 1 above)
Comment (GaelVaroquaux) [inline nilearn/decoding/fista.py]: > I recall from the ADMM version that this assignment to z was necessary there,
> too. If this operation really is in place, it wouldn't need the assignment to
> z. Does it work without it?

Actually, if z is not c-contiguous, there can be an implicit copy.

Comment (eickenberg) [inline nilearn/decoding/fista.py]: makes sense - but we have full control over z as a local variable. Why
would we let that happen? Or alternatively, shouldn't we warn about
performance loss if this situation occurs?

Comment (GaelVaroquaux) [inline nilearn/decoding/fista.py]: > makes sense - but we have full control over z as a local variable. Why
> would we let that happen? Or alternatively, shouldn't we warn about
> performance loss if this situation occurs?

I am just being extra careful. I'd rather take a small performance hit
rather than having code that might not work on a user's computer (eg
because that user uses a BLAS that I don't know).

File: nilearn/decoding/fista.py [hunk 3]
```diff
@@ -163,6 +163,9 @@ def mfista(f1_grad, f2_prox, total_energy, lipschitz_constant, w_size,
     history = []
     w_old = w.copy()
 
+    # axpy blas routine: fast y <- a*x + y
+    inplace_add_mult, = linalg.get_blas_funcs(('axpy',), (w, ))
+
```
Comment (dohmatob) [inline nilearn/decoding/fista.py]: +1.

**Comment by dohmatob** [comment]
According to circle-ci, we've lost at least 1.1% accuracy in TV-L1/haxby.

**Comment by dohmatob** [comment]
About speeding up things, maybe we could increase (but not too much, so as not to degrade stat accuracy) the .2 in the snip `dgap_factor *= .2` (line 200). Such a change is assured to lead to a linear increase in runtime.

**Comment by eickenberg** [comment]
> maybe we could increase (but not too much, so as not to degrade stat accuracy) the .2 in the code snip dgap_factor *= .2 (line 200).

+1

**Comment by eickenberg** [comment]
> According to circle-ci, we've lost at least 1.1% accuracy in TV-L1/haxby.

Interesting, that means the change in dgap tolerance schedule has changed something about the convergence of the optimizer (for better or worse on the training set, with impact on the test set)

As for numerical accuracy: Couldn't `saxpy` already do the trick? It may in general be faster to work with `float32` here.

File: nilearn/decoding/fista.py (hunk 2 above)
Comment (dohmatob) [inline nilearn/decoding/fista.py]: This looks really violent. At this point, we are 100% heuristic, and what is left of the convergence theory disappears. Will be nice to see convergence curves.

Comment (dohmatob) [inline nilearn/decoding/fista.py]: ok, it seems `dgap_tol = abs(energy_delta) / (i + 1.)` was itself a heuristic, and so no 'harm' is done. Ignore previous comment.

**Comment by GaelVaroquaux** [comment]
>     maybe we could increase (but not too much, so as not to degrade stat
>     accuracy) the .2 in the code snip dgap_factor *= .2 (line 200).

> +1

+1

**Comment by GaelVaroquaux** [comment]
> According to circle-ci, we've lost at least 1.1% accuracy in TV-L1/haxby.

Actually, I have found that run time and accuracy tend to fluctuate from
run to run. This tells me that I need to do systematic benchs and
multiple experiments. I am putting the PR on hold until I get time to do
that (I also need drago for this, ... so it's for later).

**Comment by GaelVaroquaux** [comment]
That speed up never became very reliable. Closing.
